File: examples/C/binary-tree.c
#include <stdio.h>

struct Node
{
    int value;
    struct Node *left;
    struct Node *right;
};

void depth_first_values(struct Node *node)
{
    if (node == NULL)
    {
        return;
    }

    printf("Node value: %d\n", node->value);
    depth_first_values(node->left);
    depth_first_values(node->right);
}

void breath_first_values(struct Node *node)
{
    if (node == NULL)
    {
        return;
    }

    struct Node *queue[100];
    int head = 0;
    int tail = 0;

    queue[tail++] = node;

    while (head != tail)
    {
        struct Node *current = queue[head++];

        printf("Node value: %d\n", current->value);

        if (current->left != NULL)
        {
            queue[tail++] = current->left;
        }

        if (current->right != NULL)
        {
            queue[tail++] = current->right;
        }
    }
}

int tree_includes(struct Node *node, int value)
{
    if (node == NULL)
    {
        return 0;
    }

    if (node->value == value)
    {
        return 1;
    }

    return tree_includes(node->left, value) || tree_includes(node->right, value);
}

int tree_sum(struct Node *node)
{
    if (node == NULL)
    {
        return 0;
    }

    return node->value + tree_sum(node->left) + tree_sum(node->right);
}

int tree_min_value(struct Node *node)
{
    if (node == NULL)
    {
        return 0;
    }

    int min = node->value;

    if (node->left != NULL)
    {
        int left_min = tree_min_value(node->left);

        if (left_min < min)
        {
            min = left_min;
        }
    }

    if (node->right != NULL)
    {
        int right_min = tree_min_value(node->right);

        if (right_min < min)
        {
            min = right_min;
        }
    }

    return min;
}

int max_root_to_leaf_path_sum(struct Node *node)
{
    if (node == NULL)
    {
        return 0;
    }

    int left_sum = max_root_to_leaf_path_sum(node->left);
    int right_sum = max_root_to_leaf_path_sum(node->right);

    if (left_sum > right_sum)
    {
        return node->value + left_sum;
    }
    else
    {
        return node->value + right_sum;
    }
}

int main(void)
{
    struct Node a = {1, NULL, NULL};
    struct Node b = {2, NULL, NULL};
    struct Node c = {3, NULL, NULL};
    struct Node d = {4, NULL, NULL};
    struct Node e = {5, NULL, NULL};
    struct Node f = {6, NULL, NULL};

    a.left = &b;
    a.right = &c;
    b.left = &d;
    b.right = &e;
    c.left = &f;

    printf("Node a value: %d\n", a.value);
    printf("Node b value: %d\n", a.left->value);
    printf("Node c value: %d\n", a.right->value);
    printf("Node d value: %d\n", a.left->left->value);
    printf("Node e value: %d\n", a.left->right->value);
    printf("Node f value: %d\n", a.right->left->value);

    printf("\n");
    depth_first_values(&a);

    printf("\n");
    breath_first_values(&a);

    printf("\n");
    if (tree_includes(&a, 1) == 1)
    {
        printf("Tree includes 1\n");
    }
    else
    {
        printf("Tree does not include 1\n");
    }

    if (tree_includes(&a, 9) == 1)
    {
        printf("Tree includes 9\n");
    }
    else
    {
        printf("Tree does not include 9\n");
    }

    printf("\n");
    printf("The tree sum is %d\n", tree_sum(&a));

    printf("\n");
    printf("The tree min value is %d\n", tree_min_value(&a));

    printf("\n");
    printf("The root to leaf path sum is %d\n", max_root_to_leaf_path_sum(&a));

    return 0;
}

File: examples/PHP/binary-tree.php
<?php
class Node
{
    public int $value;
    public Node|null $left = null;
    public Node|null $right = null;

    public function __construct(int $value)
    {
        $this->value = $value;
    }

    public function printValue()
    {
        print("Node a value: " . $this->value . "\n");
    }

    public function depthFirstValues()
    {
        $this->printValue();

        if ($this->left) {
            $this->left->depthFirstValues();
        }

        if ($this->right) {
            $this->right->depthFirstValues();
        }
    }

    public function breathFirstValues()
    {
        $queue = [$this];

        while (count($queue) > 0) {
            $current = array_shift($queue);
            $current->printValue();

            if ($current->left) {
                array_push($queue, $current->left);
            }

            if ($current->right) {
                array_push($queue, $current->right);
            }
        }
    }

    public function treeIncludes(int $value)
    {
        if ($this->value == $value) {
            return true;
        }

        if ($this->left) {
            return $this->left->treeIncludes($value);
        }

        if ($this->right) {
            return $this->right->treeIncludes($value);
        }

        return false;
    }

    public function treeSum()
    {
        $sum = $this->value;

        if ($this->left) {
            $sum += $this->left->treeSum();
        }

        if ($this->right) {
            $sum += $this->right->treeSum();
        }

        return $sum;
    }

    public function treeMinValue()
    {
        $min = $this->value;

        if ($this->left) {
            $min = min([$min, $this->left->treeMinValue()]);
        }

        if ($this->right) {
            $min = min([$min, $this->right->treeMinValue()]);
        }

        return $min;
    }

    public function maxRootToLeafPathSum()
    {
        $leftSum = 0;
        $rightSum = 0;

        if ($this->left) {
            $leftSum = $this->left->maxRootToLeafPathSum();
        }

        if ($this->right) {
            $rightSum = $this->right->maxRootToLeafPathSum();
        }

        if ($leftSum > $rightSum) {
            return $this->value + $leftSum;
        }

        return $this->value + $rightSum;
    }
}

$a = new Node(1);
$b = new Node(2);
$c = new Node(3);
$d = new Node(4);
$e = new Node(5);
$f = new Node(6);

$a->left = $b;
$a->right = $c;
$b->left = $d;
$b->right = $e;
$c->left = $f;

$a->printValue();
$a->left->printValue();
$a->right->printValue();
$a->left->left->printValue();
$a->left->right->printValue();
$a->right->left->printValue();

print("\n");
$a->depthFirstValues();

print("\n");
$a->breathFirstValues();

print("\n");
if ($a->treeIncludes(1)) {
    print("Tree includes 1\n");
} else {
    print("Tree does not include 1\n");
}

print("\n");
if ($a->treeIncludes(9)) {
    print("Tree includes 9\n");
} else {
    print("Tree does not include 9\n");
}

print("\n");
print("The tree sum is " . $a->treeSum() . "\n");

print("\n");
print("The tree min value is " . $a->treeMinValue() . "\n");

print("\n");
print("The root to leaf path sum is " . $a->maxRootToLeafPathSum() . "\n");